### ISSUE TITLE
pkg: orange3

### ISSUE THREAD

**Issue by InLaw** [DavHau/mach-nix#95]
```
let
   mach-nix = import (builtins.fetchGit {
     url = "https://github.com/DavHau/mach-nix/";
     ref = "refs/tags/2.2.2";
     rev = "f0b3b3c8e5153f61db7dcfbe2120c8d3e0e41371";
   }); # 

  overlays = []; # some very useful overlays
  pkgs = import mach-nix.nixpkgs.path { config = { allowUnfree = true; }; inherit overlays; };
in mach-nix.mkPythonShell rec {
  python = pkgs.python37;
  requirements =  ''   orange3   '';
}
```

```
searching for dependencies of /nix/store/g21qbhf9qf8s96ikjz5pssviszb1dlsl-python3.7-orange3-3.26.0/lib/python3.7/site-packages/Orange/projection/_som.cpython-37m-x86_64-linux-gnu.so
readelf: Error: Reading 1080 bytes extends past end of file for section headers
readelf: Error: the dynamic segment offset + size exceeds the size of the file
patchelf: section header table out of bounds
builder for '/nix/store/hmj6kndih18i59cw39nin1v4qicpfdnq-python3.7-orange3-3.26.0.drv' failed with exit code 1
cannot build derivation '/nix/store/7rvbd9ybw9548jw9mjj9sr01jar5rfgq-python3-3.7.8-env.drv': 1 dependencies couldn't be built
error: --- Error --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-shell
build of '/nix/store/7rvbd9ybw9548jw9mjj9sr01jar5rfgq-python3-3.7.8-env.drv' failed

```

**Comment by InLaw** [comment]
works on ref dev

**Comment by DavHau** [comment]
`patchelf: section header table out of bounds`
Looks like there is a problem with an elf binary in their wheel.

Maybe this should be reported upstream

**Comment by DavHau** [comment]
Reported upstream: https://github.com/biolab/orange3/issues/4956

**Comment by InLaw** [comment]
it works at current master (intentionally?)

**Comment by DavHau** [comment]
It probably worked because the problem might be specific to the wheel.


I debugged the original problem just now and I think I fixed it. The trick is to set `dontFixup = true` for this derivation.
I added this fix for orange 3 inside the new `fixes.nix` which is a global conditional override system I recently added to mach-nix (will announce more about it soon)
Please check again

**Comment by InLaw** [comment]
the installation works (but it is not running due to issues with QT ..)

**Comment by DavHau** [comment]
Which exact issues? How can i trigger them?
I noticed that there are errors during autopatchelf phase for the pyqt5 wheel. Maybe this is related?
If we can figure out the fix, we can also add this to fixes.nix

**Comment by InLaw** [comment]
install orange3

in the shell:
orange-canvas  
```
Traceback (most recent call last):
  File "/nix/store/965l4sg8sj42gxfdm5lqlr3g6q4ak1lf-python3.7-orange3-3.26.0/bin/orange-canvas", line 5, in <module>
    from Orange.canvas.__main__ import main
  File "/nix/store/m67glbj6jvjyqa625rjqyamkyj3v071q-python3-3.7.8-env/lib/python3.7/site-packages/Orange/canvas/__main__.py", line 29, in <module>
    from AnyQt.QtGui import QFont, QColor, QPalette, QDesktopServices, QIcon
  File "/nix/store/m67glbj6jvjyqa625rjqyamkyj3v071q-python3-3.7.8-env/lib/python3.7/site-packages/AnyQt/QtGui.py", line 156, in <module>
    from PySide2.QtGui import *
ImportError: libgssapi_krb5.so.2: cannot open shared object file: No such file or directory
```

**Comment by DavHau** [comment]
Sadly the PyQt5 wheel is not manylinux standard conform and comes with all kinds of unsatisfied dependencies. 
`nix-locate libgssapi_krb5.so.2` reveals that this library is found in package `nixpkgs.kerberos` for example.
You can add this to the buildInputs of the pyqt5 package like this:
```
_.pyqt5.buildInputs.add [ mach-nix.nixpkgs.kerberos ];
```

But orange-canvas seems to trigger other errors afterwards like:
`...-qtbase-5.12.7/lib/libQt5DBus.so.5: undefined symbol: _ZN14QMetaCallEventC2EttPFvP7QObjectN11QMetaObject4CallEiPPvEPKS0_iiPiS5_P10QSemaphore, version Qt_5_PRIVATE_API)`

This is probably due to incompatibilities between the libraries pyqt5 has been linked to and the ones which are available from nixpkgs. I tried a long time to fix this but wasn't able to. The only solution so far is to use an older version of pyqt5 for example  
`pyqt5 == 5.10.*`.

I added above mentioned fix to the gloabal fixes.nix. See latest commit on master: 91a263d5f40f9d75d99a5a17ffcb23aa3fbf63ba

With this version of mach-nix you can get a working version of orange-canvas like this:
```nix
let
  mach-nix = import (builtins.fetchGit {
    url = "https://github.com/DavHau/mach-nix/";
    rev = "91a263d5f40f9d75d99a5a17ffcb23aa3fbf63ba";
   });
in
with mach-nix.nixpkgs;
mach-nix.mkPython {
  requirements = ''
    orange3
    pyqt5 == 5.10.*
    sip<4.20,>=4.19.4
  '';
}
```
PS: Do you have an email address? There is no info in your github profle.

**Comment by InLaw** [comment]
this spec works for the build. Good work!
In my case the app works like you mentioned partially 
- QT is very sensitive ..

___

why is numpy build from wheels?
- I thought that it would be set by default to nipkgs?

___


if you build it with this spec there are errors like 
`QXcbConnection: XCB error: 3 (BadWindow), sequence: 54763, resource id: 41944826, major code: 40 (TranslateCoords), minor code: 0`
- when trying to give the file a name to be saved
- searching for nodes is not working (no text can be entered)


`qt.network.ssl: Incompatible version of OpenSSL`

`qt.network.ssl: QSslSocket: cannot resolve CRYPTO_num_locks`

**Comment by InLaw** [comment]
ps: updated the profile

**Comment by DavHau** [comment]
It doesn't seem like a really good idea to use binary distributions for packages as large as pyqt5 if those have such insane amount of dependencies. The chance that something will not be satisfied or is incompatible is always very high.
I can see that this is not a nix specific problem ([see here for example](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3001)).

I think such packages are meant to be built from source for the specific environment they will run in.
Its probably a better idea to start from the nixpkgs pyqt5 package and go from there. I know that also with this package something was not right for orange3. But I have much higher hopes that one could reach a stable build, fixing the problems there.

**Comment by InLaw** [comment]
you could be ~ right for orange3
but not so much for spyder (which is based on pyqt5 as well)

**Comment by DavHau** [comment]
Not sure what exactly you mean. Is there a problem with spyder?
My last statement was primarily about pyqt5, not orange3

**Comment by InLaw** [comment]
you suggested that the application is used in a isolated environment (due to it's large dependencies)
which does not hold for spyder which is as well pyqt5 dependend
(spyder does not run due to issues with pyqt5 .. )

**Comment by DavHau** [comment]
Ah, maybe I didn't express myself really well. With environment I didn't mean 'encapsulated' environment.
Let me rephrase it.
The python manylinux wheel standard intends to specify a binary format which is compatible to many linux environments. It shouldn't matter for example if you're on Debian or Arch or Centos. The manylinux standard limits the allowed set of dependencies to a particular set which is satisfied by all those systems/environments.
PyQt5's wheel is not compliant to that standard and depends on a lot more stuff than allowed. Therefore it is quite hard for any environment to fully satisfy it. This is probably why all major distros distribute a custom PyQt5 via their native package manager.

What I wanted to say is, that it is probably a bad idea to focus on getting the wheel working. Better imrpove the nixpkgs native build, to have all the features we need.

**Comment by DavHau** [comment]
I already tried `providers.pyqt5 = "nixpkgs"`. But somehow some xcb dependency is missing in the end. I'm not yet sure if this is a problem with the nixpkgs pyqt5 itself, or if this is introduced by mach-nix, due to its changes introduced by the mach-nix generated overlay.

It would help, if you could build a python environment without mach-nix to see if the same error occurs on the standard pyqt5 from nixpkgs.

**Comment by InLaw** [comment]
it works with nixos-20.03 (os-)channel 

----

with current nixpkgs-unstable 
in nix mkDerivation  and post pip install 

it fails with 
```
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb, wayland-org.kde.kwin.qpa.

```

**Comment by InLaw** [comment]
ps: spyder from nixpkgs-unstable will break my nix-shell (a channel mix is troublesome)

**Comment by DavHau** [comment]
Packges with complex dependencies like orange3 should work much better via the new conda provider.
